phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java

```diff
@@ -1039,10 +1065,20 @@ private HTableDescriptor ensureTableCreated(byte[] physicalTableName, PTableType
                 }
             }
 
+            if (isMetaTable && SchemaUtil.isNamespaceMappingEnabled(PTableType.SYSTEM,
```

Comment:
The ensureSystemTablesMigratedToSystemNamespace has race conditions in the case of multiple clients executing it at the same time. We should treat this as an upgrade step and do it while we have the SYSTEM.MUTEX lock (which ensures only a single client will be executing upgrade code).

Reply:
@JamesRTaylor We already try to acquire a lock in the SYSMUTEX table inside _ensureSystemTablesMigratedToSystemNamespace_ before we start migrating SYSTEM tables to the SYSTEM namespace, so race conditions shouldn't be a problem here, right? In case multiple clients execute this simultaneously, still only 1 would be able to migrate the tables. 

On a side-note, I still think it logically makes sense to move this to inside _upgradeSystemTables_ for consistency with the method definition.